cs/src/core/Index/FasterLog/FasterLog.cs

```diff
@@ -0,0 +1,321 @@
+            var physicalAddress = allocator.GetPhysicalAddress(logicalAddress);
+            *(int*)physicalAddress = length;
+            fixed (byte* bp = &entry.GetPinnableReference())
+                Buffer.MemoryCopy(bp, (void*)(4 + physicalAddress), length, length);
+            epoch.Suspend();
+            return logicalAddress;
+        }
+
+        /// <summary>
+        /// Append entry to log
+        /// </summary>
+        /// <param name="entry"></param>
+        /// <returns>Logical address of added entry</returns>
+        public unsafe long Append(byte[] entry)
+        {
+            epoch.Resume();
+            var length = entry.Length;
+            BlockAllocate(4 + length, out long logicalAddress);
+            var physicalAddress = allocator.GetPhysicalAddress(logicalAddress);
+            *(int*)physicalAddress = length;
+            fixed (byte* bp = entry)
+                Buffer.MemoryCopy(bp, (void*)(4 + physicalAddress), length, length);
+            epoch.Suspend();
+            return logicalAddress;
+        }
+
+        /// <summary>
+        /// Try to append entry to log
+        /// </summary>
+        /// <param name="entry">Entry to be appended to log</param>
+        /// <param name="logicalAddress">Logical address of added entry</param>
+        /// <returns>Whether the append succeeded</returns>
+        public unsafe bool TryAppend(byte[] entry, out long logicalAddress)
+        {
+            epoch.Resume();
+            logicalAddress = 0;
+            long tail = -allocator.GetTailAddress();
+            allocator.CheckForAllocateComplete(ref tail);
+            if (tail < 0)
+            { 
+                epoch.Suspend();
+                return false;
+            }
+            var length = entry.Length;
+            BlockAllocate(4 + length, out logicalAddress);
+            var physicalAddress = allocator.GetPhysicalAddress(logicalAddress);
+            *(int*)physicalAddress = length;
+            fixed (byte* bp = entry)
+                Buffer.MemoryCopy(bp, (void*)(4 + physicalAddress), length, length);
+            epoch.Suspend();
+            return true;
+        }
+
+        /// <summary>
+        /// Try to append entry to log
+        /// </summary>
+        /// <param name="entry">Entry to be appended to log</param>
+        /// <param name="logicalAddress">Logical address of added entry</param>
+        /// <returns>Whether the append succeeded</returns>
+        public unsafe bool TryAppend(Span<byte> entry, out long logicalAddress)
+        {
+            epoch.Resume();
+            logicalAddress = 0;
+            long tail = -allocator.GetTailAddress();
+            allocator.CheckForAllocateComplete(ref tail);
+            if (tail < 0)
+            {
+                epoch.Suspend();
+                return false;
+            }
+            var length = entry.Length;
+            BlockAllocate(4 + length, out logicalAddress);
```

Comment:
If there is already pending data waiting to be flushed + the incoming entry is big enough to not fit on the space remaining on the current page + no more pages can be allocated, BlockAllocate will block the thread waiting for the allocation to go through.

Is there any way for the early-exit check to account for entry.Length too?

Reply:
We only made the common path non blocking right now. Do we need it to always be non blocking? The system atomically allocates space when when it's not yet ready to use, then waits for it to become usable. If we want it to be non blocking, we will need to surface this in the API, i.e. TryAppend will return an incomplete address that the user will need to compete by calling TryCompleteAppend. Is this what you are looking for?